1. Data Exploration and Preprocessing
- Begin by loading and exploring the "sonardataset.csv" dataset. Summarize its key features such as the number of samples, features, and classes.
-	Execute necessary data preprocessing steps including data normalization, managing missing values.


In [326]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')
import keras
import tensorflow as tf

In [327]:
df = pd.read_csv('sonardataset.csv')

In [328]:
df.shape

(208, 61)

In [329]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 61 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   x_1     208 non-null    float64
 1   x_2     208 non-null    float64
 2   x_3     208 non-null    float64
 3   x_4     208 non-null    float64
 4   x_5     208 non-null    float64
 5   x_6     208 non-null    float64
 6   x_7     208 non-null    float64
 7   x_8     208 non-null    float64
 8   x_9     208 non-null    float64
 9   x_10    208 non-null    float64
 10  x_11    208 non-null    float64
 11  x_12    208 non-null    float64
 12  x_13    208 non-null    float64
 13  x_14    208 non-null    float64
 14  x_15    208 non-null    float64
 15  x_16    208 non-null    float64
 16  x_17    208 non-null    float64
 17  x_18    208 non-null    float64
 18  x_19    208 non-null    float64
 19  x_20    208 non-null    float64
 20  x_21    208 non-null    float64
 21  x_22    208 non-null    float64
 22  x_

In [330]:
df.head()

,x_1,x_2,x_3,x_4,x_5,x_6,x_7,x_8,x_9,x_10,...,x_52,x_53,x_54,x_55,x_56,x_57,x_58,x_59,x_60,Y
0,0.0200,0.0371,0.0428,0.0207,0.0954,0.0986,0.1539,0.1601,0.3109,0.2111,...,0.0027,0.0065,0.0159,0.0072,0.0167,0.0180,0.0084,0.0090,0.0032,R
1,0.0453,0.0523,0.0843,0.0689,0.1183,0.2583,0.2156,0.3481,0.3337,0.2872,...,0.0084,0.0089,0.0048,0.0094,0.0191,0.0140,0.0049,0.0052,0.0044,R
2,0.0262,0.0582,0.1099,0.1083,0.0974,0.2280,0.2431,0.3771,0.5598,0.6194,...,0.0232,0.0166,0.0095,0.0180,0.0244,0.0316,0.0164,0.0095,0.0078,R
3,0.0100,0.0171,0.0623,0.0205,0.0205,0.0368,0.1098,0.1276,0.0598,0.1264,...,0.0121,0.0036,0.0150,0.0085,0.0073,0.0050,0.0044,0.0040,0.0117,R
4,0.0762,0.0666,0.0481,0.0394,0.0590,0.0649,0.1209,0.2467,0.3564,0.4459,...,0.0031,0.0054,0.0105,0.0110,0.0015,0.0072,0.0048,0.0107,0.0094,R


In [331]:
df.tail()

,x_1,x_2,x_3,x_4,x_5,x_6,x_7,x_8,x_9,x_10,...,x_52,x_53,x_54,x_55,x_56,x_57,x_58,x_59,x_60,Y
203,0.0187,0.0346,0.0168,0.0177,0.0393,0.1630,0.2028,0.1694,0.2328,0.2684,...,0.0116,0.0098,0.0199,0.0033,0.0101,0.0065,0.0115,0.0193,0.0157,M
204,0.0323,0.0101,0.0298,0.0564,0.0760,0.0958,0.0990,0.1018,0.1030,0.2154,...,0.0061,0.0093,0.0135,0.0063,0.0063,0.0034,0.0032,0.0062,0.0067,M
205,0.0522,0.0437,0.0180,0.0292,0.0351,0.1171,0.1257,0.1178,0.1258,0.2529,...,0.0160,0.0029,0.0051,0.0062,0.0089,0.0140,0.0138,0.0077,0.0031,M
206,0.0303,0.0353,0.0490,0.0608,0.0167,0.1354,0.1465,0.1123,0.1945,0.2354,...,0.0086,0.0046,0.0126,0.0036,0.0035,0.0034,0.0079,0.0036,0.0048,M
207,0.0260,0.0363,0.0136,0.0272,0.0214,0.0338,0.0655,0.1400,0.1843,0.2354,...,0.0146,0.0129,0.0047,0.0039,0.0061,0.0040,0.0036,0.0061,0.0115,M


In [332]:
df.describe()

,x_1,x_2,x_3,x_4,x_5,x_6,x_7,x_8,x_9,x_10,...,x_51,x_52,x_53,x_54,x_55,x_56,x_57,x_58,x_59,x_60
count,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,...,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000
mean,0.029164,0.038437,0.043832,0.053892,0.075202,0.104570,0.121747,0.134799,0.178003,0.208259,...,0.016069,0.013420,0.010709,0.010941,0.009290,0.008222,0.007820,0.007949,0.007941,0.006507
std,0.022991,0.032960,0.038428,0.046528,0.055552,0.059105,0.061788,0.085152,0.118387,0.134416,...,0.012008,0.009634,0.007060,0.007301,0.007088,0.005736,0.005785,0.006470,0.006181,0.005031
min,0.001500,0.000600,0.001500,0.005800,0.006700,0.010200,0.003300,0.005500,0.007500,0.011300,...,0.000000,0.000800,0.000500,0.001000,0.000600,0.000400,0.000300,0.000300,0.000100,0.000600
25%,0.013350,0.016450,0.018950,0.024375,0.038050,0.067025,0.080900,0.080425,0.097025,0.111275,...,0.008425,0.007275,0.005075,0.005375,0.004150,0.004400,0.003700,0.003600,0.003675,0.003100
50%,0.022800,0.030800,0.034300,0.044050,0.062500,0.092150,0.106950,0.112100,0.152250,0.182400,...,0.013900,0.011400,0.009550,0.009300,0.007500,0.006850,0.005950,0.005800,0.006400,0.005300
75%,0.035550,0.047950,0.057950,0.064500,0.100275,0.134125,0.154000,0.169600,0.233425,0.268700,...,0.020825,0.016725,0.014900,0.014500,0.012100,0.010575,0.010425,0.010350,0.010325,0.008525
max,0.137100,0.233900,0.305900,0.426400,0.401000,0.382300,0.372900,0.459000,0.682800,0.710600,...,0.100400,0.070900,0.039000,0.035200,0.044700,0.039400,0.035500,0.044000,0.036400,0.043900


In [333]:
df.describe(include='object')

,Y
count,208
unique,2
top,M
freq,111


In [334]:
df.isnull().sum()

,0
x_1,0
x_2,0
x_3,0
x_4,0
x_5,0
...,...
x_57,0
x_58,0
x_59,0
x_60,0


There is no missing values

In [335]:
df.duplicated().sum()

np.int64(0)

There is no duplicate values

In [336]:
df['Y'].unique()

array(['R', 'M'], dtype=object)

In [337]:
target = df['Y']
features = df.drop('Y',axis=1)

In [338]:
features.head()

,x_1,x_2,x_3,x_4,x_5,x_6,x_7,x_8,x_9,x_10,...,x_51,x_52,x_53,x_54,x_55,x_56,x_57,x_58,x_59,x_60
0,0.0200,0.0371,0.0428,0.0207,0.0954,0.0986,0.1539,0.1601,0.3109,0.2111,...,0.0232,0.0027,0.0065,0.0159,0.0072,0.0167,0.0180,0.0084,0.0090,0.0032
1,0.0453,0.0523,0.0843,0.0689,0.1183,0.2583,0.2156,0.3481,0.3337,0.2872,...,0.0125,0.0084,0.0089,0.0048,0.0094,0.0191,0.0140,0.0049,0.0052,0.0044
2,0.0262,0.0582,0.1099,0.1083,0.0974,0.2280,0.2431,0.3771,0.5598,0.6194,...,0.0033,0.0232,0.0166,0.0095,0.0180,0.0244,0.0316,0.0164,0.0095,0.0078
3,0.0100,0.0171,0.0623,0.0205,0.0205,0.0368,0.1098,0.1276,0.0598,0.1264,...,0.0241,0.0121,0.0036,0.0150,0.0085,0.0073,0.0050,0.0044,0.0040,0.0117
4,0.0762,0.0666,0.0481,0.0394,0.0590,0.0649,0.1209,0.2467,0.3564,0.4459,...,0.0156,0.0031,0.0054,0.0105,0.0110,0.0015,0.0072,0.0048,0.0107,0.0094


In [339]:
target.head()

,Y
0,R
1,R
2,R
3,R
4,R


In [340]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder

In [341]:
x_train,x_test,y_train,y_test = train_test_split(features,target,train_size=0.8,random_state=100)
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(166, 60)
(42, 60)
(166,)
(42,)


In [342]:
## Scaling x_train and x_test data
std_sca = StandardScaler()
x_train = pd.DataFrame(std_sca.fit_transform(x_train))
x_test = pd.DataFrame(std_sca.transform(x_test))


In [343]:
x_train.head()

,0,1,2,3,4,5,6,7,8,9,...,50,51,52,53,54,55,56,57,58,59
0,-0.348907,-0.362625,-0.804504,0.508695,0.563503,1.039274,2.891795,2.593567,1.335515,1.516122,...,1.347050,-0.760544,0.723666,-0.300799,4.228681,0.316689,0.821902,3.565420,0.958659,-0.818450
1,0.328401,3.858338,3.117965,1.909293,0.682357,0.427567,1.430054,1.135347,1.055681,1.227497,...,0.553137,-0.204190,-0.490032,1.381379,1.109387,-0.324619,-0.004879,0.291374,0.616776,0.082364
2,-0.409691,-0.510404,-0.964921,-0.731038,-0.778103,-0.774586,-0.361549,-0.821995,-0.944003,-1.357552,...,0.877531,0.289180,-0.448656,2.306577,1.694255,0.114171,0.281963,0.308251,-0.984674,-0.217908
3,0.198150,-0.371861,-0.085212,0.084012,-0.399933,0.075917,0.778415,0.420917,0.076262,0.431926,...,-0.240776,-0.099218,-0.931377,-0.749380,-0.420266,-0.611520,-0.561691,-0.923735,-1.272575,-0.658305
4,-0.600726,-0.180978,0.434851,1.113548,0.741783,1.127596,0.729906,-0.270269,-0.771931,-0.644869,...,-0.616391,-0.718555,-1.262386,-0.314817,0.569510,-0.476508,-1.135375,-0.974365,-0.948687,-0.037745


In [344]:
x_test.head()

,0,1,2,3,4,5,6,7,8,9,...,50,51,52,53,54,55,56,57,58,59
0,-0.943722,-1.083051,-0.884713,-1.007725,-0.596221,-0.293725,-0.073728,-0.417070,0.172726,0.178824,...,-0.565170,-0.540102,-0.490032,-0.356872,-0.840171,-0.307743,-0.696675,-0.687464,-1.146619,-0.538197
1,-0.570334,-0.430357,-0.726882,-0.078998,0.426641,0.622201,0.380641,-0.147936,-0.069739,-0.113502,...,0.553137,-0.088721,0.020273,-0.693307,-0.930151,-0.054595,-0.747295,-0.518699,-0.966680,-0.858486
2,-0.713611,-0.698208,-0.550941,-0.763211,-0.700668,-0.570138,-0.573372,-0.885609,-0.456466,-0.573083,...,-1.188349,-0.928500,-0.448656,-1.071797,-0.600226,-0.577767,-1.034137,-0.603082,0.184925,-0.117817
3,-0.822154,-0.568901,-0.713946,-0.089723,0.158320,-0.058201,0.645824,0.894349,0.798442,0.676148,...,-0.309070,0.005755,-0.462448,-0.412944,-0.225310,-0.375249,0.028867,-0.974365,-0.408872,-0.898522
4,-0.579017,-0.519641,-0.385348,0.139778,-0.520587,-0.434385,0.222178,-0.684981,-0.642442,-0.376965,...,0.262889,0.215699,-0.641745,-0.342854,0.014635,-0.746532,-0.376087,-0.349934,-0.498841,-0.798432


In [345]:
lab_enc = LabelEncoder()

In [346]:
y_train.head()

,Y
127,M
85,R
32,R
201,M
26,R


In [347]:
y_test.head()

,Y
163,M
189,M
59,R
81,R
190,M


In [348]:
y_train.ndim

1

In [349]:
y_test.ndim

1

In [350]:
## Labeling y_test and y_train
y_train = pd.DataFrame(lab_enc.fit_transform(y_train))
y_test = pd.DataFrame(lab_enc.transform(y_test))

In [351]:
y_train.head()

,0
0,0
1,1
2,1
3,0
4,1


In [352]:
y_test.head()

,0
0,0
1,0
2,1
3,1
4,0


2. Model Implementation
-	Construct a basic ANN model using your chosen high-level neural network library. Ensure your model includes at least one hidden layer.
-	Divide the dataset into training and test sets.
-	Train your model on the training set and then use it to make predictions on the test set.

In [353]:
## Model 1

In [354]:
from keras.models import Sequential
from tensorflow.keras import layers

In [355]:
## Model Building
model = keras.Sequential()
model.add(layers.Dense(32,activation='relu',input_shape=(60,)))
model.add(layers.Dense(1,activation='sigmoid'))

In [356]:
model.summary()

Model: "sequential_13"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_21 (Dense)                │ (None, 32)             │         1,952 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_22 (Dense)                │ (None, 1)              │            33 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 1,985 (7.75 KB)

 Trainable params: 1,985 (7.75 KB)

 Non-trainable params: 0 (0.00 B)

In [357]:
60*32+32, 32*1+1

(1952, 33)

In [358]:
## Model compilation
model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

In [359]:
history = model.fit(x_train,y_train,epochs=20,validation_data=(x_test,y_test))

Epoch 1/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 3s 314ms/step - accuracy: 0.4759 - loss: 0.8698 - val_accuracy: 0.5476 - val_loss: 0.6831
Epoch 2/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 20ms/step - accuracy: 0.5181 - loss: 0.7866 - val_accuracy: 0.6190 - val_loss: 0.6312
Epoch 3/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step - accuracy: 0.5783 - loss: 0.7213 - val_accuracy: 0.6667 - val_loss: 0.5903
Epoch 4/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 29ms/step - accuracy: 0.6506 - loss: 0.6649 - val_accuracy: 0.7143 - val_loss: 0.5539
Epoch 5/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step - accuracy: 0.6807 - loss: 0.6196 - val_accuracy: 0.7381 - val_loss: 0.5249
Epoch 6/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 29ms/step - accuracy: 0.7349 - loss: 0.5740 - val_accuracy: 0.8095 - val_loss: 0.5018
Epoch 7/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 28ms/step - accuracy: 0.7651 - loss: 0.5431 - val_accuracy: 0.8571 - val_loss: 0.4838
Epoch 8/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 29ms/step - accuracy: 0.7892 - loss: 0.5133 - val_accuracy: 0.8810 - val_loss: 0.4692

In [360]:
val_loss, val_accuracy = model.evaluate(x_test,y_test)
print('Validation Loss',val_loss,'\nValidation Accuracy',val_accuracy)

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 25ms/step - accuracy: 0.8571 - loss: 0.3645
Validation Loss 0.36446240544319153 
Validation Accuracy 0.8571428656578064


In [361]:
y_pred = model.predict(x_test)
y_pred

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 45ms/step


array([[0.6850459 ],
       [0.21458979],
       [0.95568573],
       [0.46447524],
       [0.32470292],
       [0.9715713 ],
       [0.30843535],
       [0.1921296 ],
       [0.19353855],
       [0.13724788],
       [0.32150012],
       [0.1357139 ],
       [0.07150038],
       [0.98399407],
       [0.7824286 ],
       [0.88916934],
       [0.42638767],
       [0.25993636],
       [0.9444747 ],
       [0.7493977 ],
       [0.00780038],
       [0.0694398 ],
       [0.2924798 ],
       [0.39227605],
       [0.7249036 ],
       [0.21263242],
       [0.38271692],
       [0.9205438 ],
       [0.7535754 ],
       [0.41006127],
       [0.30911508],
       [0.8364114 ],
       [0.05999063],
       [0.28446648],
       [0.4200202 ],
       [0.3261299 ],
       [0.43714136],
       [0.08826338],
       [0.01846372],
       [0.14615488],
       [0.6288242 ],
       [0.72965115]], dtype=float32)

In [362]:
y_pred = np.where(y_pred>0.5,1,0)
y_pred

array([[1],
       [0],
       [1],
       [0],
       [0],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [1],
       [1],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [1],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [1]])

In [363]:
y_pred.ndim

2

In [364]:
y_test.ndim

2

In [365]:
y_pred1 = y_pred.ravel()
y_test1 = y_test.values.ravel()

In [366]:
dict = {'Predictions':y_pred1,'Actual':y_test1}
comp = pd.DataFrame(dict)
comp

,Predictions,Actual
0,1,0
1,0,0
2,1,1
3,0,1
4,0,0
5,1,1
6,0,0
7,0,0
8,0,0
9,0,0


Out of 42 predictions 9 predictions are errors

3. Hyperparameter Tuning
-	Modify various hyperparameters, such as the number of hidden layers, neurons per hidden layer, activation functions, and learning rate, to observe their impact on model performance.
-	Adopt a structured approach like grid search or random search for hyperparameter tuning, documenting your methodology thoroughly.


In [367]:
## Model 2

In [368]:
## Change in Dense value and activation function
model2 = keras.Sequential()
model2.add(layers.Dense(64,activation='tanh',input_shape=(60,)))
model2.add(layers.Dense(1,activation='sigmoid'))

In [369]:
model2.summary()

Model: "sequential_14"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_23 (Dense)                │ (None, 64)             │         3,904 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_24 (Dense)                │ (None, 1)              │            65 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 3,969 (15.50 KB)

 Trainable params: 3,969 (15.50 KB)

 Non-trainable params: 0 (0.00 B)

In [370]:
model2.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

In [371]:
history = model2.fit(x_train,y_train,epochs=20,validation_data=(x_test,y_test))

Epoch 1/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 1s 47ms/step - accuracy: 0.6566 - loss: 0.6399 - val_accuracy: 0.7619 - val_loss: 0.5072
Epoch 2/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step - accuracy: 0.7349 - loss: 0.5565 - val_accuracy: 0.7619 - val_loss: 0.4844
Epoch 3/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step - accuracy: 0.7470 - loss: 0.5068 - val_accuracy: 0.7857 - val_loss: 0.4709
Epoch 4/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step - accuracy: 0.7470 - loss: 0.4724 - val_accuracy: 0.7381 - val_loss: 0.4612
Epoch 5/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 18ms/step - accuracy: 0.8012 - loss: 0.4408 - val_accuracy: 0.8333 - val_loss: 0.4515
Epoch 6/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 15ms/step - accuracy: 0.7952 - loss: 0.4189 - val_accuracy: 0.8095 - val_loss: 0.4426
Epoch 7/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step - accuracy: 0.8253 - loss: 0.3984 - val_accuracy: 0.7857 - val_loss: 0.4396
Epoch 8/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step - accuracy: 0.8253 - loss: 0.3849 - val_accuracy: 0.7857 - val_loss: 0.4371


In [372]:
val_loss2, val_accuracy2 =model2.evaluate(x_test,y_test)
print('Validation Loss',val_loss,'\nValidation Accuracy',val_accuracy2)

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step - accuracy: 0.8095 - loss: 0.4188
Validation Loss 0.36446240544319153 
Validation Accuracy 0.8095238208770752


In [373]:
y_pred2 = model2.predict(x_test)
y_pred2

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 47ms/step


array([[0.5837546 ],
       [0.29111117],
       [0.97585   ],
       [0.23418126],
       [0.52787167],
       [0.9920601 ],
       [0.18414481],
       [0.04338146],
       [0.01319281],
       [0.07815517],
       [0.7611224 ],
       [0.01725573],
       [0.330613  ],
       [0.9934347 ],
       [0.86024374],
       [0.8998497 ],
       [0.3077279 ],
       [0.3362188 ],
       [0.9800277 ],
       [0.89691854],
       [0.01527069],
       [0.0299481 ],
       [0.43607867],
       [0.1250881 ],
       [0.43967858],
       [0.20440505],
       [0.18765728],
       [0.99284065],
       [0.6647225 ],
       [0.43542275],
       [0.27445266],
       [0.96737486],
       [0.01475383],
       [0.5806629 ],
       [0.94801486],
       [0.58895737],
       [0.16990966],
       [0.03654571],
       [0.00122073],
       [0.3678505 ],
       [0.6918259 ],
       [0.53949535]], dtype=float32)

In [374]:
y_pred2 = np.where(y_pred2>0.5,1,0)
y_pred2

array([[1],
       [0],
       [1],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [1],
       [0],
       [0],
       [1],
       [1],
       [1],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [1],
       [0],
       [1],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [1],
       [1]])

In [375]:
dict2 = {'Predictions':y_pred2.ravel(),'Actual':y_test1}
comp2 = pd.DataFrame(dict2)
comp2

,Predictions,Actual
0,1,0
1,0,0
2,1,1
3,0,1
4,1,0
5,1,1
6,0,0
7,0,0
8,0,0
9,0,0


out of 41 predictions 9 errors

In [376]:
## Model 3

In [377]:
model3 =keras.Sequential()
model3.add(layers.Dense(64,activation='relu',input_shape=(60,)))
model3.add(layers.Dense(32,activation='relu',input_shape=(60,)))
model3.add(layers.Dense(1,activation='sigmoid'))

In [378]:
model3.summary()

Model: "sequential_15"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_25 (Dense)                │ (None, 64)             │         3,904 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_26 (Dense)                │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_27 (Dense)                │ (None, 1)              │            33 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 6,017 (23.50 KB)

 Trainable params: 6,017 (23.50 KB)

 Non-trainable params: 0 (0.00 B)

In [379]:
from tensorflow.keras.optimizers import Adam

In [380]:
model3.compile(optimizer=Adam(learning_rate=0.01),loss='binary_crossentropy',metrics=['accuracy'])

In [381]:
history = model3.fit(x_train,y_train,epochs=20,validation_data=(x_test,y_test))

Epoch 1/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 1s 47ms/step - accuracy: 0.6325 - loss: 0.6621 - val_accuracy: 0.7381 - val_loss: 0.4372
Epoch 2/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 15ms/step - accuracy: 0.8494 - loss: 0.3800 - val_accuracy: 0.8095 - val_loss: 0.4199
Epoch 3/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 15ms/step - accuracy: 0.9458 - loss: 0.2392 - val_accuracy: 0.8571 - val_loss: 0.3061
Epoch 4/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 17ms/step - accuracy: 0.9458 - loss: 0.1396 - val_accuracy: 0.8571 - val_loss: 0.3120
Epoch 5/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 15ms/step - accuracy: 1.0000 - loss: 0.0658 - val_accuracy: 0.9048 - val_loss: 0.3390
Epoch 6/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 15ms/step - accuracy: 1.0000 - loss: 0.0320 - val_accuracy: 0.9048 - val_loss: 0.3383
Epoch 7/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 18ms/step - accuracy: 1.0000 - loss: 0.0153 - val_accuracy: 0.9048 - val_loss: 0.3395
Epoch 8/20
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 15ms/step - accuracy: 1.0000 - loss: 0.0083 - val_accuracy: 0.8810 - val_loss: 0.3660


In [382]:
val_loss2, val_accuracy2 =model2.evaluate(x_test,y_test)
print('Validation Loss',val_loss,'\nValidation Accuracy',val_accuracy2)

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 24ms/step - accuracy: 0.8095 - loss: 0.4188
Validation Loss 0.36446240544319153 
Validation Accuracy 0.8095238208770752


In [383]:
y_pred3 = model3.predict(x_test)
y_pred3 = np.where(y_pred3>0.5,1,0)
y_pred3

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 50ms/step


array([[1],
       [0],
       [1],
       [1],
       [0],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [1],
       [1],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [1],
       [0],
       [0],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [1]])

In [384]:
dict3 = {'Predictions':y_pred3.ravel(),'Actual':y_test1}
comp3 = pd.DataFrame(dict3)
comp3

,Predictions,Actual
0,1,0
1,0,0
2,1,1
3,1,1
4,0,0
5,1,1
6,0,0
7,0,0
8,0,0
9,0,0


In [385]:
## GridSearchCv

In [386]:
param_grid = {
    'neurons': [16, 32, 64],
    'activation': ['relu', 'tanh'],
    'learning_rate': [0.01, 0.001],
    'batch_size': [8, 16]
}

In [388]:
def build_model(neurons=32, activation='relu', learning_rate=0.001):
    model4 = keras.Sequential([
        layers.Dense(neurons, activation=activation, input_shape=(60,)),
        layers.Dense(1, activation='sigmoid')
    ])

    model4.compile(
        optimizer=Adam(learning_rate=learning_rate),
        loss='binary_crossentropy',
        metrics=['accuracy']
    )

    return model4

In [389]:
results = []

for neurons in [16, 32, 64]:
    for activation in ['relu', 'tanh']:
        for lr in [0.01, 0.001]:

            model4 = build_model(neurons, activation, lr)

            history = model4.fit(
                x_train, y_train,
                epochs=20,
                batch_size=8,
                verbose=0
            )

            loss, acc = model4.evaluate(x_test, y_test, verbose=0)

            results.append({
                'neurons': neurons,
                'activation': activation,
                'learning_rate': lr,
                'accuracy': acc
            })

In [393]:
y_pred4 = model4.predict(x_test)
y_pred4 = np.where(y_pred4>0.5,1,0)
y_pred4

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 85ms/step


array([[1],
       [0],
       [1],
       [1],
       [0],
       [1],
       [0],
       [0],
       [0],
       [0],
       [1],
       [0],
       [0],
       [1],
       [1],
       [1],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [1],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [1],
       [1]])

In [390]:
results_df = pd.DataFrame(results)
print(results_df.sort_values(by='accuracy', ascending=False))

    neurons activation  learning_rate  accuracy
9        64       relu          0.001  0.904762
10       64       tanh          0.010  0.904762
0        16       relu          0.010  0.880952
5        32       relu          0.001  0.880952
11       64       tanh          0.001  0.880952
8        64       relu          0.010  0.880952
1        16       relu          0.001  0.857143
2        16       tanh          0.010  0.857143
3        16       tanh          0.001  0.857143
4        32       relu          0.010  0.857143
6        32       tanh          0.010  0.857143
7        32       tanh          0.001  0.785714


In [395]:
dict4 = {'Predictions':y_pred4.ravel(),'Actual':y_test1}
comp4 = pd.DataFrame(dict4)
comp4

,Predictions,Actual
0,1,0
1,0,0
2,1,1
3,1,1
4,0,0
5,1,1
6,0,0
7,0,0
8,0,0
9,0,0


In [391]:
from sklearn.metrics import classification_report

4. Evaluation
-	Employ suitable metrics such as accuracy, precision, recall, and F1-score to evaluate your model's performance.
-	Discuss the performance differences between the model with default hyperparameters and the tuned model, emphasizing the effects of hyperparameter tuning.

In [397]:
## Model 1
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.89      0.89      0.89        28
           1       0.79      0.79      0.79        14

    accuracy                           0.86        42
   macro avg       0.84      0.84      0.84        42
weighted avg       0.86      0.86      0.86        42



In [398]:
## Model 2
print(classification_report(y_test,y_pred2))

              precision    recall  f1-score   support

           0       0.92      0.79      0.85        28
           1       0.67      0.86      0.75        14

    accuracy                           0.81        42
   macro avg       0.79      0.82      0.80        42
weighted avg       0.83      0.81      0.81        42



In [399]:
## Model 3
print(classification_report(y_test,y_pred3))

              precision    recall  f1-score   support

           0       0.93      0.89      0.91        28
           1       0.80      0.86      0.83        14

    accuracy                           0.88        42
   macro avg       0.86      0.88      0.87        42
weighted avg       0.88      0.88      0.88        42



In [401]:
## Model 4
print(classification_report(y_test,y_pred4))

              precision    recall  f1-score   support

           0       0.96      0.86      0.91        28
           1       0.76      0.93      0.84        14

    accuracy                           0.88        42
   macro avg       0.86      0.89      0.87        42
weighted avg       0.89      0.88      0.88        42



- The 1st model achieved good balanced performance with 86% accuracy
- Increasing neurons and using tanh degraded performance of model 2 due to instability
- Adding an additional hidden layer improved accuracy to 88% and provided better class balance
- Grid search tuning further improved recall for the minority class (Mine), making the model more sensitive to detecting mines
- Overall, Model 3 provided the best balance, while Model 4 is preferable when minimizing false negatives is critical